channeldb/migration29/migration.go

```diff
@@ -0,0 +1,498 @@
+	var total uint64
+
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if openChanBucket == nil {
+		return 0, fmt.Errorf("root bucket not found")
+	}
+
+	// Locate the next migration height.
+	locator, err := locateNextUpdateNum(openChanBucket)
+	if err != nil {
+		return 0, fmt.Errorf("locator got error: %v", err)
+	}
+
+	// If the returned locator is nil, we've done migrating the logs.
+	if locator == nil {
+		return 0, nil
+	}
+
+	// cb is the callback function to be used when iterating the buckets.
+	cb := func(chanBucket kvdb.RwBucket, l *updateLocator) error {
+		// Read the sub-bucket level 4.
+		logBucket := chanBucket.NestedReadBucket(
+			revocationLogBucketDeprecated,
+		)
+		// Exit early if we don't have the old bucket.
+		if logBucket == nil {
+			return nil
+		}
+
+		// Jump to the end of the cursor.
+		key, _ := logBucket.ReadCursor().Last()
+
+		// Since the CommitHeight is a zero-based monotonically
+		// increased index, its value plus one reflects the total
+		// records under this chan bucket.
+		lastHeight := binary.BigEndian.Uint64(key) + 1
+
+		total += lastHeight
+		return nil
+	}
+
+	err = iterateBuckets(openChanBucket, locator, cb)
+	return total, err
+}
+
+// logEntry houses the info needed to write a new revocation log.
+type logEntry struct {
+	log          *RevocationLog
+	commitHeight uint64
+	ourIndex     uint32
+	theirIndex   uint32
+	locator      *updateLocator
+}
+
+// logEntries maps a bucket locator to a list of entries under that bucket.
+type logEntries map[*updateLocator][]*logEntry
+
+// result is made of two channels that's used to send back the constructed new
+// revocation log or an error.
+type result struct {
+	newLog  chan *logEntry
+	errChan chan error
+}
+
+// readOldRevocationLogs finds a list of old revocation logs and converts them
+// into the new revocation logs.
+func readOldRevocationLogs(openChanBucket kvdb.RwBucket,
+	locator *updateLocator) (logEntries, error) {
+
+	entries := make(logEntries)
+	results := make([]*result, 0)
+
+	var wg sync.WaitGroup
+
+	// collectLogs is a helper closure that reads all newly created
+	// revocation logs sent over the result channels.
+	//
+	// NOTE: the order of the logs cannot be guaranteed, which is fine as
+	// boltdb will take care of the orders when saving them.
+	collectLogs := func() error {
+		wg.Wait()
+
+		for _, r := range results {
+			select {
+			case entry := <-r.newLog:
+				entries[entry.locator] = append(
+					entries[entry.locator], entry,
+				)
+
+			case err := <-r.errChan:
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// createLog is a helper closure that constructs a new revocation log.
+	//
+	// NOTE: used as a goroutine.
+	createLog := func(chanState *mig26.OpenChannel,
+		c mig.ChannelCommitment, l *updateLocator, r *result) {
+
+		defer wg.Done()
+
+		// Find the output indexes.
+		ourIndex, theirIndex, err := findOutputIndexes(chanState, &c)
```

Comment:
Was running a migration, the first 67% or so was pretty quick, but then I noticed it really start to get bogged down. I took a profile, and realized that this is the function that soaks up the most CPU: 
![Screen Shot 2022-06-30 at 7 39 34 PM](https://user-images.githubusercontent.com/998190/176812943-ee8e8c41-f708-499b-a41b-63d15bcd29da.png)

Reply:
One idea here would be to make a goroutine worker pool and submit this job to them.